include/fiddle/interaction/interaction_utilities.h

```diff
@@ -260,6 +263,155 @@ namespace fdl
+                           dealii::Point<2>        r,
+                           dealii::Tensor<1, 2>    q,
+                           const double            tol);
+
+  /**
+   * This code is used to determine if the given line would intersect with the
+   * triangle Also perturb r if the FD stencil cutting boundary of the triangle.
+   *
+   * https://mathworld.wolfram.com/TriangleInterior.html
+   *
+   * @param[in] elements_vertices The array containing Eulerian location of
+   * vertices of the given element.
+   *
+   * @param[in] r The center location of the FD stencil`.
+   *
+   * @param[in] q a unit vector aligned with the coordinate axes.
+   *
+   * @param[in] tol a given tolerance for deciding if the intersection point is
+   * within the edge.
+   */
+  bool
+  if_line_intersect_with_triangle(
+    const std::array<Point<3>, 3> &element_vetices,
+    dealii::Point<3>              &r,
+    const dealii::Tensor<1, 3>    &q);
+
+  /**
+   * This code is used to determine if the given line would intersect with a
+   * rectangle containing the triangle
+   *
+   * @param[in] elements_vertices The array containing Eulerian location of
+   * vertices of the given element.
+   *
+   * @param[in] r The center location of the FD stencil`.
+   *
+   * @param[in] q a unit vector aligned with the coordinate axes.
+   *
+   * @param[in] tol a given tolerance for deciding if the intersection point is
+   * within the edge.
+   */
+  bool
+  if_line_intersect_with_box(const std::array<Point<3>, 3> &element_vertices,
+                             dealii::Point<3>              &r,
+                             const dealii::Tensor<1, 3>    &q);
+
+  /**
+   * Compute intersection the point of a line with a flat triangle
+   *
+   * WARNING: This code is specialized to the case in which q is a unit vector
+   * aligned with the coordinate axes.
+   *
+   * Linear interpolation:
+   *
+   * (1-u-v)*p0 + u*p1 + v*p2 = p + t*d
+   *
+   * https://en.wikipedia.org/wiki/M%C3%B6ller%E2%80%93Trumbore_intersection_algorithm
+   *
+   * @param[out] t_vals The vector containing the intersection point in Eulerian
+   * coordinate and element reference coordinate.
+   *
+   * @param[in] element_vertices The array containing Eulerian location of
+   * vertices of the given element.
+   *
+   * @param[in] r The center location of the FD stencil`.
+   *
+   * @param[in] q a unit vector aligned with the coordinate axes.
+   *
+   * @param[in] tol a given tolerance for deciding if the intersection point is
+   * within the edge.
+   *
+   * @note This function is only work for P1 element.
+   */
+  void
+  intersect_line_with_flat_triangle(
+    std::vector<std::pair<double, Point<2>>> &t_vals,
+    std::array<Point<3>, 3>                   element_vetices,
+    dealii::Point<3>                          r,
+    const dealii::Tensor<1, 3>                q,
+    const double                              tol = 0);
```

Comment:
Lets combine this with the other function and let the implementation switch between spacedim = 2 and spacedim = 3.